Yosemite/Yosemite/Tools/POS/POSOrderService.swift

```diff
@@ -126,7 +131,15 @@ public final class POSOrderService: POSOrderServiceProtocol {
         } else {
             syncedOrder = try await ordersRemote.createPOSOrder(siteID: siteID, order: order, fields: [.items, .status])
         }
-        return POSOrder(order: syncedOrder)
+
+        return try await withCheckedThrowingContinuation { continuation in
+            // Upserting synced order to storage is necessary because `OrderAction.retrieveOrder` dispatched in `CollectOrderPaymentUseCase`
+            // returns the order in storage after the first time.
+            // Otherwise, the order total could be outdated for payment collection if it differs from the first synced order.
```

Comment:
Is there any way you can think of that we could fix this underlying issue, so we can avoid upserting the order and all the potential issues it could cause?

Reply:
The issue is from `CollectOrderPaymentUseCase` dispatching `OrderAction.retrieveOrder` to either load the order from remote (for the first time) or storage (after the first time): 

https://github.com/woocommerce/woocommerce-ios/blob/4b4d37e579477b02943453462fb13b5b86ee5c57/WooCommerce/Classes/ViewRelated/Orders/Collect%20Payments/CollectOrderPaymentUseCase.swift#L250

We could also rely on the `Order` passed to `CollectOrderPaymentUseCase` to check on the amount & paid status:

https://github.com/woocommerce/woocommerce-ios/blob/4b4d37e579477b02943453462fb13b5b86ee5c57/WooCommerce/Classes/ViewRelated/Orders/Collect%20Payments/CollectOrderPaymentUseCase.swift#L263-L269

Like by adding a parameter to `CollectOrderPaymentUseCase.collectPayment` like `shouldRetrieveOrderBeforeValidation: Bool` so that it's default to `true` for the existing app and `false` for POS (and maybe future use cases) to skip this Yosemite action and use the DI'ed order. This way, only a small amount of changes are in the existing app code and no changes are expected in the existing app flow. WDYT of this approach? I imagine there was a reason `OrderAction.retrieveOrder` is needed in `CollectOrderPaymentUseCase` in the first place.

Reply:
The call during the payment flow is intended to make sure that we're working with the most up to date version of the order possible, in case it's been paid from some other source.

That's more relevant to the order list... however, if it's getting the order from _storage_ the second time, it's not really doing the job properly. We check because we want to make sure we're not using a copy of an order that's a few days out of date from Storage.

I don't think we should use the DI'ed Order, but we could change the app call to always get it from the network.

Reply:
> The call during the payment flow is intended to make sure that we're working with the most up to date version of the order possible, in case it's been paid from some other source.

In this case, always loading the order remotely feels like the expected behavior. Two ways I can think of:
- Add a parameter `OrderAction.retrieveOrder` like `alwaysFromRemote: Bool` to always load the order from the network for the payment collection use case
- Create a separate action like `OrderAction.retrieveOrderRemotely` to always sync the order from the network
- As a performance optimization that `OrderAction.retrieveOrder` has internally, create a separate `OrderAction` like `loadOrderForPaymentValidation` that only loads the order's `total` and `datePaid` remotely which are the 2 fields used for checking if an order is valid for payment

Because this requires changing the existing app behavior, I'll create an issue after your confirmation.

Reply:
> As a performance optimization that OrderAction.retrieveOrder has internally, create a separate OrderAction like loadOrderForPaymentValidation that only loads the order's total and datePaid remotely which are the 2 fields used for checking if an order is valid for payment

This one sounds nice to me, and even better it might be `checkPaymentStatus(of: Order) -> PaymentStatus`. That will need more changes to the logic though, so let's leave the optimisation for the future. There are lots of things which could be better about our decisions on whether orders are paid or not!

> Create a separate action like OrderAction.retrieveOrderRemotely to always sync the order from the network

Let's do this one 😊 it's clear, simple, and doesn't involve changing the payments code significantly.

Reply:
Thanks for the quick confirmation! I added a subtask to replace the upsert logic with this new action in https://github.com/woocommerce/woocommerce-ios/issues/12869, will work on this after checking the order update API (can also hand this off to Bozidar after time-boxing the effort since it's more on order sync than payment).